go/caddy/pfpki/models.go

```diff
@@ -0,0 +1,755 @@
+	Information.NextCursor = vars.Cursor + vars.Limit
+	if vars.Cursor < count {
+		sql, err := vars.Sql(c)
+		if err != nil {
+			Information.Error = err.Error()
+			return Information, errors.New("A database error occured. See log for details.")
+		}
+		var cadb []CA
+		pfpki.DB.Select(sql.Select).Order(sql.Order).Offset(sql.Offset).Limit(sql.Limit).Find(&cadb)
+		Information.Entries = cadb
+	}
+
+	return Information, nil
+}
+
+func (c CA) search(pfpki *Handler, vars Vars) (Info, error) {
+	Information := Info{}
+	sql, err := vars.Sql(c)
+	if err != nil {
+		Information.Error = err.Error()
+		return Information, errors.New("A database error occured. See log for details.")
+	}
+	var count int
+	pfpki.DB.Model(&CA{}).Where(sql.Where.Query, sql.Where.Values...).Count(&count)
+	Information.TotalCount = count
+	Information.PrevCursor = vars.Cursor
+	Information.NextCursor = vars.Cursor + vars.Limit
+	if vars.Cursor < count {
+		var cadb []CA
+		pfpki.DB.Select(sql.Select).Where(sql.Where.Query, sql.Where.Values...).Order(sql.Order).Offset(sql.Offset).Limit(sql.Limit).Find(&cadb)
+		Information.Entries = cadb
+	}
+
+	return Information, nil
+}
+
+func (p Profile) new(pfpki *Handler) (Info, error) {
+
+	var profiledb []Profile
+	var err error
+	Information := Info{}
+	switch *p.KeyType {
+	case KEY_RSA:
+		if p.KeySize < 2048 {
+			err = errors.New("invalid private key size, should be at least 2048")
+			Information.Error = err.Error()
+			return Information, err
+		}
+	case KEY_ECDSA:
+		if !(p.KeySize == 256 || p.KeySize == 384 || p.KeySize == 521) {
+			err = errors.New("invalid private key size, should be 256 or 384 or 521")
+			Information.Error = err.Error()
+			return Information, err
+		}
+	case KEY_DSA:
+		if !(p.KeySize == 1024 || p.KeySize == 2048 || p.KeySize == 3072) {
+			err = errors.New("invalid private key size, should be 1024 or 2048 or 3072")
+			Information.Error = err.Error()
+			return Information, err
+		}
+	default:
+		return Information, errors.New("KeyType unsupported")
+
+	}
+
+	var ca CA
+	if CaDB := pfpki.DB.First(&ca, p.CaID).Find(&ca); CaDB.Error != nil {
+		Information.Error = err.Error()
+		return Information, CaDB.Error
+	}
+
+	if err := pfpki.DB.Create(&Profile{Name: p.Name, Ca: ca, CaID: p.CaID, CaName: ca.Cn, Validity: p.Validity, KeyType: p.KeyType, KeySize: p.KeySize, Digest: p.Digest, KeyUsage: p.KeyUsage, ExtendedKeyUsage: p.ExtendedKeyUsage, P12MailPassword: p.P12MailPassword, P12MailSubject: p.P12MailSubject, P12MailFrom: p.P12MailFrom, P12MailHeader: p.P12MailHeader, P12MailFooter: p.P12MailFooter}).Error; err != nil {
+		Information.Error = err.Error()
+		return Information, errors.New("A database error occured. See log for details.")
+	}
+	pfpki.DB.Select("id, name, ca_id, ca_name, validity, key_type, key_size, digest, key_usage, extended_key_usage, p12_mail_password, p12_mail_subject, p12_mail_from, p12_mail_header, p12_mail_footer").Where("name = ?", p.Name).First(&profiledb)
```

Comment:
I don't think you have to specify all the columns in gorm when performing a select

Reply:
It's to get rid of some columns